google/cloud/forseti/services/inventory/storage.py

```diff
@@ -638,17 +598,40 @@ def _get_resource_rows(self, key):
             Exception: if there is no such row or more than one.
         """
 
-        qry = (
-            self.session.query(Inventory)
-            .filter(Inventory.inventory_index_id == self.inventory_index.id)
-            .filter(Inventory.resource_id == key))
-        rows = qry.all()
+        rows = self.session.query(Inventory).filter(
+            and_(
+                Inventory.inventory_index_id == self.inventory_index.id,
+                Inventory.resource_id == key,
+                Inventory.resource_type == resource_type
+            )).all()
 
         if not rows:
             raise Exception('Resource {} not found in the table'.format(key))
         else:
             return rows
 
+    def _get_resource_id(self, resource):
+        """Checks if a resource exists already in the inventory.
+
+        Args:
+            resource (object): Resource object to check against the db.
+
+        Returns:
+            int: The resource id of the existing resource, else 0.
+        """
+        row = self.session.query(Inventory.id).filter(
+            and_(
+                Inventory.inventory_index_id == self.inventory_index.id,
+                Inventory.category == Categories.resource,
+                Inventory.resource_type == resource.type(),
+                Inventory.resource_id == resource.key(),
+            )).one_or_none()
```

Comment:
one_or_none() will raise `sqlalchemy.orm.exc.MultipleResultsFound` if multiple object identities are returned. We are expecting multiple rows with same inventory_index_id, category, resource_type and resource_id, what should we do in that case?

Reply:
This code is written to ensure that there is only ever one or none. In write, if one already exists the write is transformed to an update to ensure there are no duplicates.

If we ever get a MultipleResultsFound that should be an exception to blocks the creation of the inventory because a core assertion has been violated.

Reply:
Instead of query the database for every single insert, what do you think about having a look up table to increase the performance? if the size of the table is a concern maybe we can use a customized bloom filter?